R/insert-ui.R

```diff
@@ -172,3 +172,275 @@ removeUI <- function(selector,
   if (!immediate) session$onFlushed(callback, once = TRUE)
   else callback()
 }
+
+
+#' Dynamically add a tabPanel
+#'
+#' Dynamically add a \code{tabPanel()} into an existing \code{tabsetPanel}.
+#'
+#' @param tabsetPanelId The \code{id} of the \code{tabsetPanel()} into which
+#'   \code{tab} will be inserted.
+#'
+#' @param tab The tab element to be added (must be created with
+#'   \code{tabPanel}).
+#'
+#' @param target The \code{value} of an existing \code{tabPanel()}, next to
+#'   which \code{tab} will be added. If \code{NULL},the \code{tab} will be
+#'   placed either as the first tab or the last tab, depending on the
+#'   \code{position} argument.
```

Comment:
I can't decide the `target=NULL` semantics or not. You've got it like this:

* If target is non-`NULL`, `right` means "to the right of the target".
* If target is `NULL`, `right` means "to the right of everything".

This overloading of the meaning of `position` feels strange to me. It also means that there are two ways of putting a tab at the end using `right`: either use a `NULL` target, or use the current last tab as the target; but, there is no way to put a tab at the beginning, without changing the position to `left`.

The usual way I've seen this done is:

* If target is non-`NULL`, `right` means "to the right of the target". (no change)
* If target is `NULL`, `right` means _"to the right of nothing"_--that is, the leftmost position.

This way, the meaning of `position` is more consistent. And also both `position="right"` and `"left"` are capable of addressing any possible slot.